swarm/storage/pyramid.go

```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+			index:         0,
+			updatePending: true,
+		}
+		copy(newEntry.chunk[8:], chunk.Key)
+		chunkLevel[0] = append(chunkLevel[0], newEntry)
+		return nil
+	}
+
+	var treeSize int64
+	var depth int
+	treeSize = self.chunkSize
+	for ; treeSize < chunk.Size; treeSize *= self.branches {
+		depth++
+	}
+
+	// Add the root chunk entry
+	branchCount := int64(len(chunk.SData)-8) / self.hashSize
+	newEntry := &TreeEntry{
+		level:         int(depth - 1),
+		branchCount:   branchCount,
+		subtreeSize:   uint64(chunk.Size),
+		chunk:         chunk.SData,
+		key:           key,
+		index:         0,
+		updatePending: true,
+	}
+	chunkLevel[depth-1] = append(chunkLevel[depth-1], newEntry)
+
+	// Add the rest of the tree
+	for lvl := (depth - 1); lvl >= 1; lvl-- {
+
+		//TODO(jmozah): instead of loading finished branches and then trim in the end,
+		//avoid loading them in the first place
+		for _, ent := range chunkLevel[lvl] {
+			branchCount = int64(len(ent.chunk)-8) / self.hashSize
+			for i := int64(0); i < branchCount; i++ {
+				key := ent.chunk[8+(i*self.hashSize) : 8+((i+1)*self.hashSize)]
+				newChunk := retrieve(key, chunkC, quitC)
+				if newChunk == nil {
+					return errLoadingTreeChunk
 				}
-			}
-			hash := hasher.Sum(nil)
-			last := task.Last || (node != nil) && node.Last
-			// Insert the subresult into the memoization tree
-			results.Lock.Lock()
-			if node = results.Levels[depth][task.Index/pow]; node == nil {
-				// Figure out the pending tasks
-				pending := self.branches
-				if task.Index/pow == results.Chunks/pow {
-					pending = (results.Chunks + pow/self.branches - 1) / (pow / self.branches) % self.branches
+				bewBranchCount := int64(len(newChunk.SData)-8) / self.hashSize
+				newEntry := &TreeEntry{
+					level:         int(lvl - 1),
+					branchCount:   bewBranchCount,
+					subtreeSize:   uint64(newChunk.Size),
+					chunk:         newChunk.SData,
+					key:           key,
+					index:         0,
+					updatePending: true,
 				}
-				node = &Node{pending, 0, make([]common.Hash, pending), last}
-				results.Levels[depth][task.Index/pow] = node
+				chunkLevel[lvl-1] = append(chunkLevel[lvl-1], newEntry)
+
 			}
-			node.Pending--
-			i := task.Index / (pow / self.branches) % self.branches
-			if last {
-				node.Last = true
+
+			// We need to get only the right most unfinished branch.. so trim all finished branches
+			if int64(len(chunkLevel[lvl-1])) >= self.branches {
+				chunkLevel[lvl-1] = nil
 			}
-			copy(node.Children[i][:], hash)
-			node.Size += size
-			left := node.Pending
-			if chunkC != nil {
-				if swg != nil {
-					swg.Add(1)
-				}
+		}
+	}
+
+	return nil
+}
+
+func (self *PyramidChunker) prepareChunks(isAppend bool, chunkLevel [][]*TreeEntry, data io.Reader, rootKey []byte, quitC chan bool, wg *sync.WaitGroup, jobC chan *chunkJob, processorWG *sync.WaitGroup, chunkC chan *Chunk, errC chan error, storageWG *sync.WaitGroup) {
+	defer wg.Done()
+
+	chunkWG := &sync.WaitGroup{}
```

Comment:
What is this wait group used for?

Reply:
This is where the parent  node waits for all the children nodes (128 branches) to complete OR for the input to get exhausted. Only when all the 128 hashes of the child nodes are processed.. the parent chunk can be constructed.